scripts/cxg_remaster.py

```diff
@@ -0,0 +1,155 @@
+import argparse
+import time
+
+import numpy as np
+import psutil
+import tiledb
+
+"""
+Re-master an existing "X" tiledb array in a CXG
+Creates a new array with the same data but a different tdb schema
+"""
+
+X_extent = None
+X_name = None
+
+
+def main():
+    parser = argparse.ArgumentParser()
+    parser.add_argument("cxg", type=str, help="cxg path")
+    parser.add_argument("--kind", choices=["sparse", "dense"], type=str, help="target: sparse or dense")
+    parser.add_argument("--obs", type=int, help="obs extent", default=512)
+    parser.add_argument("--var", type=int, help="var extent", default=2048)
+    parser.add_argument("--cell-order", choices=["row", "col"], type=str, default="row")
+    parser.add_argument("--tile-order", choices=["row", "col"], type=str, default="col")
+    parser.add_argument("--capacity", type=int, default=128000)
+    parser.add_argument("--compression", type=int, default=22)
+    parser.add_argument("--name", type=str, default="X_new")
+    parser.add_argument("--arr", type=str, default="X")
+    args = parser.parse_args()
+
+    global X_extent, X_name
+    X_extent = [args.obs, args.var]
+    X_name = args.name
+    out_sparse = args.kind == "sparse"
+
+    ctx = create_fast_ctx(
+        {
+            "py.init_buffer_bytes": 32 * 1024 ** 3,
+            "py.exact_init_buffer_bytes": "true",
+            "sm.tile_cache_size": 2 * 1024 ** 3,
+        }
+    )
+    st = time.time()
+    with tiledb.scope_ctx(ctx):
+        with tiledb.open(f"{args.cxg}/{args.arr}", "r") as old_X:
+            create_new_X(args, old_X.schema)
+            print("created, starting to read...")
+            in_sparse = old_X.schema.sparse
+            with tiledb.open(f"{args.cxg}/{X_name}", "w") as new_X:
+                if in_sparse and out_sparse:
+                    print("sparse->sparse")
+                    i, chunk = 0, 200_000
+                    while i < old_X.shape[0]:
+                        print(f"Chunk {i}")
+                        dat = old_X[i:i + chunk]
+                        print(f"Got dat")
+                        new_X[dat["obs"], dat["var"]] = dat[""]
+                        i += chunk
+                elif in_sparse and not out_sparse:
+                    raise ValueError("spare->dense not supported")
+                elif not in_sparse and out_sparse:
+                    print("dense->sparse")
+                    i, chunk = 0, 20_000
+                    while i < old_X.shape[0]:
+                        print(f"Chunk {i}")
+                        dat = old_X[i:i+chunk]
+                        datnz = np.nonzero(dat)
+                        new_X[datnz[0] + i, datnz[1]] = dat[datnz[0], datnz[1]]
+                        i += chunk
+                else:
+                    print("dense->dense")
+                    new_X[:] = old_X[:]
+
+        print(f"Took {time.time() - st} s")
+
+    print("done.")
+
+
+def create_new_X(args, old_schema):
```

Comment:
Most of this code duplicates the code in cxg_generation_utils.py that create new arrays.  Is it worth having a single function, and importing it?  Just concerned about these diverging....

Reply:
Agree there's a divergence risk, but I prefer to keep them separate since this is intended more as a one-off script